### PR TITLE
Fix 'Illegal instant due to time zone offset transition'

### PR DESCRIPTION
**Pull request by wendigo**
## Description

Fixes https://github.com/trinodb/trino/issues/6242

## Release notes

( ) This is not user-visible or docs only and no release notes are required.
( ) Release notes are required, please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: client/trino-jdbc/src/test/java/io/trino/jdbc/BaseTestJdbcResultSet.java [hunk 1]
```diff
@@ -283,24 +283,12 @@ public void testDate()
...
-                // TODO (https://github.com/trinodb/trino/issues/6242) this should not fail
-                assertThatThrownBy(() -> rs.getObject(column, Date.class))
-                        .isInstanceOf(SQLException.class)
-                        .hasMessage("Expected value to be a date but is: 1970-01-01")
-                        .hasStackTraceContaining("Cannot parse \"1970-01-01\": Illegal instant due to time zone offset transition (America/Bahia_Banderas)");
+                assertEquals(rs.getObject(column), sqlDate);
+                assertEquals(rs.getObject(column, Date.class), sqlDate);
                 // TODO assertEquals(rs.getObject(column, LocalDate.class), localDate);
-
-                // TODO (https://github.com/trinodb/trino/issues/6242) this should not fail
```
Comment (kokosing) [inline client/trino-jdbc/src/test/java/io/trino/jdbc/BaseTestJdbcResultSet.java:299]: Are you fixing https://github.com/trinodb/trino/issues/6242?

File: client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java [hunk 2]
```diff
@@ -361,9 +353,9 @@ private Date getDate(int columnIndex, DateTimeZone localTimeZone)
 
     private static Date parseDate(String value, DateTimeZone localTimeZone)
     {
-        long millis = DATE_FORMATTER.withZone(localTimeZone).parseMillis(String.valueOf(value));
-        if (millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND) {
-            return new Date(millis);
+        Optional<Date> parsedDate = tryParseDate(value, localTimeZone);
```
Comment (kokosing) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java:357]: inline `tryParseDate`

File: client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java [hunk 3]
```diff
@@ -380,6 +372,23 @@ private static Date parseDate(String value, DateTimeZone localTimeZone)
...
+    {
+        try {
+            long millis = DATE_FORMATTER.withZone(localTimeZone).parseMillis(String.valueOf(value));
+            if (millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND) {
+                return Optional.of(new Date(millis));
+            }
+        }
+        catch (IllegalArgumentException ignored) {
+            // https://joda-time.sourceforge.net/faq.html#illegalinstant
+            LocalDate localDate = DATE_FORMATTER.parseLocalDate(String.valueOf(value));
```
Comment (kokosing) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java:386]: Why not to use this always?

Comment (wendigo) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java]: It won't work for every date (pre-gregorian ones)

File: client/trino-jdbc/src/test/java/io/trino/jdbc/BaseTestJdbcResultSet.java (hunk 1 above)
Comment (wendigo) [reply client/trino-jdbc/src/test/java/io/trino/jdbc/BaseTestJdbcResultSet.java:299]: Yes.

File: client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java [hunk 4]
```diff
@@ -361,9 +354,16 @@ private Date getDate(int columnIndex, DateTimeZone localTimeZone)
...
-        if (millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND) {
-            return new Date(millis);
+        try {
+            long millis = DATE_FORMATTER.withZone(localTimeZone).parseMillis(String.valueOf(value));
+            if (millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND) {
+                return new Date(millis);
+            }
+        }
+        catch (IllegalArgumentException ignored) {
+            // https://joda-time.sourceforge.net/faq.html#illegalinstant
```
Comment (kokosing) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java:364]: Can we somehow validate the exception. IllegalArgumentException looks generic.

**Review by kokosing — APPROVED** [review]
% comment

File: client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java (hunk 4 above)
Comment (wendigo) [reply client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java:365]: I've used `IllegalInstantException` instead

**Review by kokosing — APPROVED** [review]
@electrum would you like to take a look?

File: client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java [hunk 5]
```diff
@@ -361,9 +355,16 @@ private Date getDate(int columnIndex, DateTimeZone localTimeZone)
...
+        try {
+            long millis = DATE_FORMATTER.withZone(localTimeZone).parseMillis(String.valueOf(value));
+            if (millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND) {
+                return new Date(millis);
+            }
+        }
+        catch (IllegalInstantException ignored) {
+            // https://joda-time.sourceforge.net/faq.html#illegalinstant
+            LocalDate localDate = DATE_FORMATTER.parseLocalDate(String.valueOf(value));
+            return new Date(localDate.toDateTimeAtStartOfDay(localTimeZone).getMillis());
```
Comment (findepi) [inline client/trino-jdbc/src/main/java/io/trino/jdbc/AbstractTrinoResultSet.java:367]: - shouldn't this be done only if `millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND`?
 the code block below is supposed to be handling `!(millis >= START_OF_MODERN_ERA_SECONDS * MILLISECONDS_PER_SECOND)` cases, but it's no longer the case

- why do we parse with two different ways? can't we use `DATE_FORMATTER.parseLocalDate` always?
  if there is a good reason for this, there should be a code comment explaining it

**Pull request merged**